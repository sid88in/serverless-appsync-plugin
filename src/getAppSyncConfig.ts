import {
  AppSyncConfig,
  isSharedApiConfig,
  PipelineResolverConfig,
  UnitResolverConfig,
} from './types/index.js';

import type {
  ApiKeyConfig,
  AppSyncConfig as PluginAppSyncConfig,
  DataSourceConfig,
  PipelineFunctionConfig,
  ResolverConfig,
  BaseAppSyncConfig,
  SharedAppSyncConfig,
  FullAppSyncConfig,
  Substitutions,
} from './types/plugin.js';
import { forEach, merge } from 'lodash-es';

const flattenMaps = <T>(
  input?: Record<string, T> | Record<string, T>[],
): Record<string, T> => {
  if (Array.isArray(input)) {
    return merge({}, ...input);
  } else {
    return merge({}, input);
  }
};

export const isUnitResolver = (resolver: {
  kind?: 'UNIT' | 'PIPELINE';
}): resolver is UnitResolverConfig => {
  return resolver.kind === 'UNIT';
};

export const isPipelineResolver = (resolver: {
  kind?: 'UNIT' | 'PIPELINE';
}): resolver is PipelineResolverConfig => {
  return !resolver.kind || resolver.kind === 'PIPELINE';
};

const toResourceName = (name: string) => {
  return name.replace(/[^a-z_]/i, '_');
};

export const getAppSyncConfig = (
  config: AppSyncConfig,
): PluginAppSyncConfig => {
  const baseConfig = getBaseAppsyncConfig(config);

  // handle shared appsync config
  if (isSharedApiConfig(config)) {
    const apiId: string = config.apiId;
    return {
      ...baseConfig,
      apiId,
    } satisfies SharedAppSyncConfig;
  }

  // Handle full appsync config
  const schema = Array.isArray(config.schema)
    ? config.schema
    : [config.schema || 'schema.graphql'];

  const additionalAuthentications = config.additionalAuthentications || [];

  let apiKeys: Record<string, ApiKeyConfig> | undefined;
  if (
    config.authentication?.type === 'API_KEY' ||
    additionalAuthentications.some((auth) => auth.type === 'API_KEY')
  ) {
    const inputKeys = config.apiKeys || [];

    apiKeys = inputKeys.reduce((acc, key) => {
      if (typeof key === 'string') {
        acc[key] = { name: key };
      } else {
        acc[key.name] = key;
      }

      return acc;
    }, {});
  }

  return {
    ...config,
    ...baseConfig,
    additionalAuthentications,
    apiKeys,
    schema,
  } satisfies FullAppSyncConfig;
};

function getBaseAppsyncConfig(config: AppSyncConfig): BaseAppSyncConfig {
  const dataSources: Record<string, DataSourceConfig> = {};
  const resolvers: Record<string, ResolverConfig> = {};
  const pipelineFunctions: Record<string, PipelineFunctionConfig> = {};
  const substitutions: Substitutions = {};

  forEach(flattenMaps(config.dataSources), (ds, name) => {
    dataSources[name] = {
      ...ds,
      name,
    };
  });

  forEach(flattenMaps(config.resolvers), (resolver, typeAndField) => {
    const [type, field] = typeAndField.split('.');

    if (typeof resolver === 'string') {
      resolvers[typeAndField] = {
        dataSource: resolver,
        kind: 'UNIT',
        type,
        field,
      };
      return;
    }

    if (isUnitResolver(resolver) && typeof resolver.dataSource === 'object') {
      const name = typeAndField.replace(/[^a-z_]/i, '_');
      dataSources[name] = {
        ...resolver.dataSource,
        name,
      };
    }

    resolvers[typeAndField] = {
      ...resolver,
      type: resolver.type || type,
      field: resolver.field || field,
      ...(isUnitResolver(resolver)
        ? {
            kind: 'UNIT',
            dataSource:
              typeof resolver.dataSource === 'object'
                ? typeAndField.replace(/[^a-z_]/i, '_')
                : resolver.dataSource,
          }
        : {
            kind: 'PIPELINE',
            functions: resolver.functions.map((f, index) => {
              if (typeof f === 'string') {
                return f;
              }

              const name = `${toResourceName(typeAndField)}_${index}`;
              pipelineFunctions[name] = {
                ...f,
                name,
                dataSource:
                  typeof f.dataSource === 'string' ? f.dataSource : name,
              };
              if (typeof f.dataSource === 'object') {
                dataSources[name] = {
                  ...f.dataSource,
                  name,
                };
              }
              return name;
            }),
          }),
    };
  });

  forEach(flattenMaps(config.pipelineFunctions), (func, name) => {
    if (typeof func.dataSource === 'object') {
      dataSources[name] = {
        ...func.dataSource,
        name,
      };
    }

    pipelineFunctions[name] = {
      ...func,
      dataSource: typeof func.dataSource === 'string' ? func.dataSource : name,
      name,
    };
  });

  if (config.substitutions) {
    Object.assign(substitutions, config.substitutions);
  }

  return {
    dataSources,
    resolvers,
    pipelineFunctions,
    substitutions,
  };
}
