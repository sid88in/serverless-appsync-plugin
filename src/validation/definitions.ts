export const stringOrIntrinsicFunction = {
  oneOf: [
    { type: 'string' },
    {
      type: 'object',
      required: [],
      additionalProperties: true,
    },
  ],
  errorMessage: 'must be a string or a CloudFormation intrinsic function',
};
// Depends on stringOrIntrinsicFunction
export const substitutions = {
  type: 'object',
  additionalProperties: {
    $ref: '#/definitions/stringOrIntrinsicFunction',
  },
  required: [],
  errorMessage: 'must be a valid substitutions definition',
};

export const AUTH_TYPES = [
  'AMAZON_COGNITO_USER_POOLS',
  'AWS_LAMBDA',
  'OPENID_CONNECT',
  'AWS_IAM',
  'API_KEY',
] as const;

export const DATASOURCE_TYPES = [
  'AMAZON_DYNAMODB',
  'AMAZON_OPENSEARCH_SERVICE',
  'AWS_LAMBDA',
  'HTTP',
  'NONE',
  'RELATIONAL_DATABASE',
  'AMAZON_EVENTBRIDGE',
] as const;

// Depends on stringOrIntrinsicFunction
export const lambdaFunctionConfig = {
  oneOf: [
    {
      type: 'object',
      properties: {
        functionName: { type: 'string' },
        functionAlias: { type: 'string' },
      },
      required: ['functionName'],
    },
    {
      type: 'object',
      properties: {
        functionArn: {
          $ref: '#/definitions/stringOrIntrinsicFunction',
        },
      },
      required: ['functionArn'],
    },
    {
      type: 'object',
      properties: {
        function: { type: 'object' },
      },
      required: ['function'],
    },
  ],
  errorMessage:
    'must specify functionName, functionArn or function (all exclusives)',
};

//depends on cognitoAuth, lambdaAuth and oidcAuth
export const auth = {
  type: 'object',
  title: 'Authentication',
  description: 'Authentication type and definition',
  properties: {
    type: {
      type: 'string',
      enum: AUTH_TYPES,
      errorMessage: `must be one of ${AUTH_TYPES.join(', ')}`,
    },
  },
  if: { properties: { type: { const: 'AMAZON_COGNITO_USER_POOLS' } } },
  then: {
    properties: { config: { $ref: '#/definitions/cognitoAuth' } },
    required: ['config'],
  },
  else: {
    if: { properties: { type: { const: 'AWS_LAMBDA' } } },
    then: {
      properties: { config: { $ref: '#/definitions/lambdaAuth' } },
      required: ['config'],
    },
    else: {
      if: { properties: { type: { const: 'OPENID_CONNECT' } } },
      then: {
        properties: { config: { $ref: '#/definitions/oidcAuth' } },
        required: ['config'],
      },
    },
  },
  required: ['type'],
};

// Depends on stringOrIntrinsicFunction
export const cognitoAuth = {
  type: 'object',
  properties: {
    userPoolId: { $ref: '#/definitions/stringOrIntrinsicFunction' },
    awsRegion: { $ref: '#/definitions/stringOrIntrinsicFunction' },
    defaultAction: {
      type: 'string',
      enum: ['ALLOW', 'DENY'],
      errorMessage: 'must be "ALLOW" or "DENY"',
    },
    appIdClientRegex: { $ref: '#/definitions/stringOrIntrinsicFunction' },
  },
  required: ['userPoolId'],
};

// Depends on lambdaFunctionConfig
export const lambdaAuth = {
  type: 'object',
  oneOf: [{ $ref: '#/definitions/lambdaFunctionConfig' }],
  properties: {
    // Note: functionName and functionArn are already defined in #/definitions/lambdaFunctionConfig
    // But if not also defined here, TypeScript shows an error.
    functionName: { type: 'string' },
    functionArn: { type: 'string' },
    identityValidationExpression: { type: 'string' },
    authorizerResultTtlInSeconds: { type: 'number' },
  },
  required: [],
};

export const oidcAuth = {
  type: 'object',
  properties: {
    issuer: { type: 'string' },
    clientId: { type: 'string' },
    iatTTL: { type: 'number' },
    authTTL: { type: 'number' },
  },
  required: ['issuer'],
};

export const iamAuth = {
  type: 'object',
  properties: {
    type: {
      type: 'string',
      const: 'AWS_IAM',
    },
  },
  required: ['type'],
  errorMessage: 'must be a valid IAM config',
};

export const apiKeyAuth = {
  type: 'object',
  properties: {
    type: {
      type: 'string',
      const: 'API_KEY',
    },
  },
  required: ['type'],
  errorMessage: 'must be a valid API_KEY config',
};

export const visibilityConfig = {
  type: 'object',
  properties: {
    cloudWatchMetricsEnabled: { type: 'boolean' },
    name: { type: 'string' },
    sampledRequestsEnabled: { type: 'boolean' },
  },
  required: [],
};

// Depends on visibilityConfig and customWafRule
export const wafRule = {
  anyOf: [
    { type: 'string', enum: ['throttle', 'disableIntrospection'] },
    {
      type: 'object',
      properties: {
        disableIntrospection: {
          type: 'object',
          properties: {
            name: { type: 'string' },
            priority: { type: 'integer' },
            visibilityConfig: { $ref: '#/definitions/visibilityConfig' },
          },
        },
      },
      required: ['disableIntrospection'],
    },
    {
      type: 'object',
      properties: {
        throttle: {
          oneOf: [
            { type: 'integer', minimum: 100 },
            {
              type: 'object',
              properties: {
                name: { type: 'string' },
                action: {
                  type: 'string',
                  enum: ['Allow', 'Block'],
                },
                aggregateKeyType: {
                  type: 'string',
                  enum: ['IP', 'FORWARDED_IP'],
                },
                limit: { type: 'integer', minimum: 100 },
                priority: { type: 'integer' },
                scopeDownStatement: { type: 'object' },
                forwardedIPConfig: {
                  type: 'object',
                  properties: {
                    headerName: {
                      type: 'string',
                      pattern: '^[a-zA-Z0-9-]+$',
                    },
                    fallbackBehavior: {
                      type: 'string',
                      enum: ['MATCH', 'NO_MATCH'],
                    },
                  },
                  required: ['headerName', 'fallbackBehavior'],
                },
                visibilityConfig: {
                  $ref: '#/definitions/visibilityConfig',
                },
              },
              required: [],
            },
          ],
        },
      },
      required: ['throttle'],
    },
    { $ref: '#/definitions/customWafRule' },
  ],
  errorMessage: 'must be a valid WAF rule',
};
// Depends on visibilityConfig
export const customWafRule = {
  type: 'object',
  properties: {
    name: { type: 'string' },
    priority: { type: 'number' },
    action: {
      type: 'string',
      enum: ['Allow', 'Block', 'Count', 'Captcha'],
    },
    statement: { type: 'object', required: [] },
    visibilityConfig: { $ref: '#/definitions/visibilityConfig' },
  },
  required: ['name', 'statement'],
};

// Depends on stringOrIntrinsicFunction
export const environment = {
  type: 'object',
  additionalProperties: {
    $ref: '#/definitions/stringOrIntrinsicFunction',
  },
  required: [],
  errorMessage: 'must be a valid environment definition',
};
// Depends on dataSourceConfig
export const dataSource = {
  if: { type: 'object' },
  then: { $ref: '#/definitions/dataSourceConfig' },
  else: {
    type: 'string',
    errorMessage: 'must be a string or data source definition',
  },
};
// Depends on substitutions, resolverCachingConfig, dataSource, pipelineFunction
export const resolverConfig = {
  type: 'object',
  properties: {
    kind: {
      type: 'string',
      enum: ['UNIT', 'PIPELINE'],
      errorMessage: 'must be "UNIT" or "PIPELINE"',
    },
    type: { type: 'string' },
    field: { type: 'string' },
    maxBatchSize: { type: 'number', minimum: 1, maximum: 2000 },
    code: { type: 'string' },
    request: { type: 'string' },
    response: { type: 'string' },
    sync: { $ref: '#/definitions/syncConfig' },
    substitutions: { $ref: '#/definitions/substitutions' },
    caching: { $ref: '#/definitions/resolverCachingConfig' },
  },
  if: { properties: { kind: { const: 'UNIT' } }, required: ['kind'] },
  then: {
    properties: {
      dataSource: { $ref: '#/definitions/dataSource' },
    },
    required: ['dataSource'],
  },
  else: {
    properties: {
      functions: {
        type: 'array',
        items: { $ref: '#/definitions/pipelineFunction' },
      },
    },
    required: ['functions'],
  },
  required: [],
};
// Depends on resolverConfig
export const resolverConfigMap = {
  type: 'object',
  patternProperties: {
    // Type.field keys, type and field are not required
    '^[_A-Za-z][_0-9A-Za-z]*\\.[_A-Za-z][_0-9A-Za-z]*$': {
      $ref: '#/definitions/resolverConfig',
    },
  },
  additionalProperties: {
    // Other keys, type and field are required
    $merge: {
      source: { $ref: '#/definitions/resolverConfig' },
      with: { required: ['type', 'field'] },
    },
    errorMessage: {
      required: {
        type: 'resolver definitions that do not specify Type.field in the key must specify the type and field as properties',
        field:
          'resolver definitions that do not specify Type.field in the key must specify the type and field as properties',
      },
    },
  },
  required: [],
};
// Depends on dataSource
export const pipelineFunctionConfig = {
  type: 'object',
  properties: {
    dataSource: { $ref: '#/definitions/dataSource' },
    description: { type: 'string' },
    request: { type: 'string' },
    response: { type: 'string' },
    sync: { $ref: '#/definitions/syncConfig' },
    maxBatchSize: { type: 'number', minimum: 1, maximum: 2000 },
    substitutions: { $ref: '#/definitions/substitutions' },
  },
  required: ['dataSource'],
};
// Depends on pipelineFunctionConfig
export const pipelineFunction = {
  if: { type: 'object' },
  then: { $ref: '#/definitions/pipelineFunctionConfig' },
  else: {
    type: 'string',
    errorMessage: 'must be a string or function definition',
  },
};
// Depends on pipelineFunctionConfig
export const pipelineFunctionConfigMap = {
  type: 'object',
  additionalProperties: {
    if: { type: 'object' },
    then: { $ref: '#/definitions/pipelineFunctionConfig' },
    else: {
      type: 'string',
      errorMessage: 'must be a string or an object',
    },
  },
  required: [],
};
export const resolverCachingConfig = {
  oneOf: [
    { type: 'boolean' },
    {
      type: 'object',
      properties: {
        ttl: { type: 'integer', minimum: 1, maximum: 3600 },
        keys: {
          type: 'array',
          items: { type: 'string' },
        },
      },
      required: [],
    },
  ],
  errorMessage: 'must be a valid resolver caching config',
};
// Depends on lambdaFunctionConfig
export const syncConfig = {
  type: 'object',
  if: { properties: { conflictHandler: { const: ['LAMBDA'] } } },
  then: { $ref: '#/definitions/lambdaFunctionConfig' },
  properties: {
    functionArn: { type: 'string' },
    functionName: { type: 'string' },
    conflictDetection: { type: 'string', enum: ['VERSION', 'NONE'] },
    conflictHandler: {
      type: 'string',
      enum: ['LAMBDA', 'OPTIMISTIC_CONCURRENCY', 'AUTOMERGE'],
    },
  },
  required: [],
};
// Depends on stringOrIntrinsicFunction
export const iamRoleStatements = {
  type: 'array',
  items: {
    type: 'object',
    properties: {
      Effect: { type: 'string', enum: ['Allow', 'Deny'] },
      Action: { type: 'array', items: { type: 'string' } },
      Resource: {
        oneOf: [
          { $ref: '#/definitions/stringOrIntrinsicFunction' },
          {
            type: 'array',
            items: { $ref: '#/definitions/stringOrIntrinsicFunction' },
          },
        ],
        errorMessage: 'contains invalid resolver definitions',
      },
    },
    required: ['Effect', 'Action', 'Resource'],
    errorMessage: 'must be a valid IAM role statement',
  },
};
// Depends on dataSourceDynamoDb, datasourceLambdaConfig, dataSourceHttpConfig, datasourceEsConfig, datasourceRelationalDbConfig and datasourceEventBridgeConfig
export const dataSourceConfig = {
  type: 'object',
  properties: {
    type: {
      type: 'string',
      enum: DATASOURCE_TYPES,
      errorMessage: `must be one of ${DATASOURCE_TYPES.join(', ')}`,
    },
    description: { type: 'string' },
  },
  if: { properties: { type: { const: 'AMAZON_DYNAMODB' } } },
  then: {
    properties: { config: { $ref: '#/definitions/dataSourceDynamoDb' } },
    required: ['config'],
  },
  else: {
    if: { properties: { type: { const: 'AWS_LAMBDA' } } },
    then: {
      properties: {
        config: { $ref: '#/definitions/datasourceLambdaConfig' },
      },
      required: ['config'],
    },
    else: {
      if: { properties: { type: { const: 'HTTP' } } },
      then: {
        properties: {
          config: { $ref: '#/definitions/dataSourceHttpConfig' },
        },
        required: ['config'],
      },
      else: {
        if: {
          properties: {
            type: { const: 'AMAZON_OPENSEARCH_SERVICE' },
          },
        },
        then: {
          properties: {
            config: { $ref: '#/definitions/datasourceEsConfig' },
          },
          required: ['config'],
        },
        else: {
          if: { properties: { type: { const: 'RELATIONAL_DATABASE' } } },
          then: {
            properties: {
              config: {
                $ref: '#/definitions/datasourceRelationalDbConfig',
              },
            },
            required: ['config'],
          },
          else: {
            if: { properties: { type: { const: 'AMAZON_EVENTBRIDGE' } } },
            then: {
              properties: {
                config: {
                  $ref: '#/definitions/datasourceEventBridgeConfig',
                },
              },
              required: ['config'],
            },
          },
        },
      },
    },
  },
  required: ['type'],
};
// Depends on stringOrIntrinsicFunction and iamRoleStatements
export const dataSourceHttpConfig = {
  type: 'object',
  properties: {
    endpoint: { $ref: '#/definitions/stringOrIntrinsicFunction' },
    serviceRoleArn: {
      $ref: '#/definitions/stringOrIntrinsicFunction',
    },
    iamRoleStatements: {
      $ref: '#/definitions/iamRoleStatements',
    },
    authorizationConfig: {
      type: 'object',
      properties: {
        authorizationType: {
          type: 'string',
          enum: ['AWS_IAM'],
          errorMessage: 'must be AWS_IAM',
        },
        awsIamConfig: {
          type: 'object',
          properties: {
            signingRegion: {
              $ref: '#/definitions/stringOrIntrinsicFunction',
            },
            signingServiceName: {
              $ref: '#/definitions/stringOrIntrinsicFunction',
            },
          },
          required: ['signingRegion'],
        },
      },
      required: ['authorizationType', 'awsIamConfig'],
    },
  },
  required: ['endpoint'],
};
// Depends on stringOrIntrinsicFunction and iamRoleStatements
export const dataSourceDynamoDb = {
  type: 'object',
  properties: {
    tableName: { $ref: '#/definitions/stringOrIntrinsicFunction' },
    useCallerCredentials: { type: 'boolean' },
    serviceRoleArn: {
      $ref: '#/definitions/stringOrIntrinsicFunction',
    },
    region: {
      $ref: '#/definitions/stringOrIntrinsicFunction',
    },
    iamRoleStatements: {
      $ref: '#/definitions/iamRoleStatements',
    },
    versioned: { type: 'boolean' },
    deltaSyncConfig: {
      type: 'object',
      properties: {
        deltaSyncTableName: { type: 'string' },
        baseTableTTL: { type: 'integer' },
        deltaSyncTableTTL: { type: 'integer' },
      },
      required: ['deltaSyncTableName'],
    },
  },
  required: ['tableName'],
};
// Depends on stringOrIntrinsicFunction and iamRoleStatements
export const datasourceRelationalDbConfig = {
  type: 'object',
  properties: {
    region: { $ref: '#/definitions/stringOrIntrinsicFunction' },
    relationalDatabaseSourceType: {
      type: 'string',
      enum: ['RDS_HTTP_ENDPOINT'],
    },
    serviceRoleArn: { $ref: '#/definitions/stringOrIntrinsicFunction' },
    dbClusterIdentifier: {
      $ref: '#/definitions/stringOrIntrinsicFunction',
    },
    databaseName: { $ref: '#/definitions/stringOrIntrinsicFunction' },
    schema: { type: 'string' },
    awsSecretStoreArn: {
      $ref: '#/definitions/stringOrIntrinsicFunction',
    },
    iamRoleStatements: {
      $ref: '#/definitions/iamRoleStatements',
    },
  },
  required: ['awsSecretStoreArn', 'dbClusterIdentifier'],
};
// Depends on lambdaFunctionConfig, stringOrIntrinsicFunction and iamRoleStatements
export const datasourceLambdaConfig = {
  type: 'object',
  oneOf: [
    {
      $ref: '#/definitions/lambdaFunctionConfig',
    },
  ],
  properties: {
    functionName: { type: 'string' },
    functionArn: {
      $ref: '#/definitions/stringOrIntrinsicFunction',
    },
    serviceRoleArn: { $ref: '#/definitions/stringOrIntrinsicFunction' },
    iamRoleStatements: { $ref: '#/definitions/iamRoleStatements' },
  },
  required: [],
};
// Depends on stringOrIntrinsicFunction and iamRoleStatements
export const datasourceEsConfig = {
  type: 'object',
  oneOf: [
    {
      oneOf: [
        {
          type: 'object',
          properties: {
            endpoint: { $ref: '#/definitions/stringOrIntrinsicFunction' },
          },
          required: ['endpoint'],
        },
        {
          type: 'object',
          properties: {
            domain: { $ref: '#/definitions/stringOrIntrinsicFunction' },
          },
          required: ['domain'],
        },
      ],
      errorMessage: 'must have a endpoint or domain (but not both)',
    },
  ],
  properties: {
    endpoint: { $ref: '#/definitions/stringOrIntrinsicFunction' },
    domain: { $ref: '#/definitions/stringOrIntrinsicFunction' },
    region: { $ref: '#/definitions/stringOrIntrinsicFunction' },
    serviceRoleArn: { $ref: '#/definitions/stringOrIntrinsicFunction' },
    iamRoleStatements: { $ref: '#/definitions/iamRoleStatements' },
  },
  required: [],
};
// Depends on stringOrIntrinsicFunction
export const datasourceEventBridgeConfig = {
  type: 'object',
  properties: {
    eventBusArn: { $ref: '#/definitions/stringOrIntrinsicFunction' },
  },
  required: ['eventBusArn'],
};
