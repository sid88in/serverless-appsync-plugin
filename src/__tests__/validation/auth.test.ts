import { AppSyncConfigInput } from '../../getAppSyncConfig';
import { validateConfig } from '../../validation';
import { basicConfig } from '../basicConfig';

describe('Valdiation', () => {
  describe('Valid', () => {
    const assertions = [
      {
        name: 'Api Key',
        config: {
          ...basicConfig,
          authentication: {
            type: 'API_KEY',
          },
        } as AppSyncConfigInput,
      },
      {
        name: 'Cognito',
        config: {
          ...basicConfig,
          authentication: {
            type: 'AMAZON_COGNITO_USER_POOLS',
            config: {
              userPoolId: '123456',
              awsRegion: 'us-east-1',
              defaultAction: 'ALLOW',
              appIdClientRegex: '.*',
            },
          },
        } as AppSyncConfigInput,
      },
      {
        name: 'Cognito with Refs',
        config: {
          ...basicConfig,
          authentication: {
            type: 'AMAZON_COGNITO_USER_POOLS',
            config: {
              userPoolId: {
                Ref: 'CognitoUserPool',
              },
              appIdClientRegex: {
                Ref: 'CognitoUserPoolClient',
              },
            },
          },
        } as AppSyncConfigInput,
      },
      {
        name: 'OIDC',
        config: {
          ...basicConfig,
          authentication: {
            type: 'OPENID_CONNECT',
            config: {
              issuer: 'https://auth.example.com',
              clientId: '90941906-004b-4cc5-9685-6864a8e08835',
              iatTTL: 3600,
              authTTL: 3600,
            },
          },
        } as AppSyncConfigInput,
      },
      {
        name: 'IAM',
        config: {
          ...basicConfig,
          authentication: {
            type: 'AWS_IAM',
          },
        } as AppSyncConfigInput,
      },
      {
        name: 'Lambda with functionName',
        config: {
          ...basicConfig,
          authentication: {
            type: 'AWS_LAMBDA',
            config: {
              functionName: 'myFunction',
              identityValidationExpression: '*',
              authorizerResultTtlInSeconds: 600,
            },
          },
        } as AppSyncConfigInput,
      },
      {
        name: 'Lambda with functionArn',
        config: {
          ...basicConfig,
          authentication: {
            type: 'AWS_LAMBDA',
            config: {
              functionArn: 'arn:aws:lambda:...',
            },
          },
        } as AppSyncConfigInput,
      },
    ];

    assertions.forEach((config) => {
      it(`should validate a ${config.name}`, () => {
        expect(validateConfig(config.config)).toBe(true);
      });
    });
  });

  describe('Invalid', () => {
    const assertions = [
      {
        name: 'Cognito missing config',
        config: {
          ...basicConfig,
          authentication: {
            type: 'AMAZON_COGNITO_USER_POOLS',
          },
        },
      },
      {
        name: 'Cognito empty config',
        config: {
          ...basicConfig,
          authentication: {
            type: 'AMAZON_COGNITO_USER_POOLS',
            config: {},
          },
        },
      },
      {
        name: 'Cognito with invalid userPoolId',
        config: {
          ...basicConfig,
          authentication: {
            type: 'AMAZON_COGNITO_USER_POOLS',
            config: {
              userPoolId: 124,
              awsRegion: 456,
              defaultAction: 'Foo',
              appIdClientRegex: 123,
            },
          },
        },
      },
      {
        name: 'OIDC with missing config',
        config: {
          ...basicConfig,
          authentication: {
            type: 'OPENID_CONNECT',
          },
        },
      },
      {
        name: 'OIDC with empty config',
        config: {
          ...basicConfig,
          authentication: {
            type: 'OPENID_CONNECT',
            config: {},
          },
        },
      },
      {
        name: 'OIDC with invalid config',
        config: {
          ...basicConfig,
          authentication: {
            type: 'OPENID_CONNECT',
            config: {
              issuer: 123,
              clientId: 456,
              iatTTL: 'foo',
              authTTL: 'bar',
            },
          },
        },
      },
      {
        name: 'Lambda with missing config',
        config: {
          ...basicConfig,
          authentication: {
            type: 'AWS_LAMBDA',
          },
        },
      },
      {
        name: 'Lambda with empty config',
        config: {
          ...basicConfig,
          authentication: {
            type: 'AWS_LAMBDA',
            config: {},
          },
        },
      },
      {
        name: 'Lambda with invalid functionName and functionVersion',
        config: {
          ...basicConfig,
          authentication: {
            type: 'AWS_LAMBDA',
            config: {
              functionName: 123,
              functionVersion: 123,
              identityValidationExpression: 456,
              authorizerResultTtlInSeconds: 'foo',
            },
          },
        },
      },
      {
        name: 'Lambda with invalid config functionnArn',
        config: {
          ...basicConfig,
          authentication: {
            type: 'AWS_LAMBDA',
            config: {
              functionArn: 123,
              identityValidationExpression: 456,
              authorizerResultTtlInSeconds: 'foo',
            },
          },
        },
      },
      {
        name: 'Lambda with invalid config: both functionName and functionnArn are set',
        config: {
          ...basicConfig,
          authentication: {
            type: 'AWS_LAMBDA',
            config: {
              functionName: 'myFunction',
              functionArn: 'arn:lambda:',
            },
          },
        },
      },
    ];

    assertions.forEach((config) => {
      it(`should validate a ${config.name}`, () => {
        expect(function () {
          validateConfig(config.config);
        }).toThrowErrorMatchingSnapshot();
      });
    });
  });
});
