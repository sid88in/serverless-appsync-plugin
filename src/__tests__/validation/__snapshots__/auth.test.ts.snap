// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Valdiation Invalid should validate a Cognito empty config 1`] = `"/authentication/config: must have required property 'userPoolId'"`;

exports[`Valdiation Invalid should validate a Cognito missing config 1`] = `"/authentication: must have required property 'config'"`;

exports[`Valdiation Invalid should validate a Cognito with invalid userPoolId 1`] = `
"/authentication/config/userPoolId: must be a string or a CloudFormation intrinsic function
/authentication/config/awsRegion: must be a string or a CloudFormation intrinsic function
/authentication/config/defaultAction: must be \\"ALLOW\\" or \\"DENY\\"
/authentication/config/appIdClientRegex: must be a string or a CloudFormation intrinsic function"
`;

exports[`Valdiation Invalid should validate a Lambda with empty config 1`] = `"/authentication/config: must specify functionName, functionArn or function (all exclusives)"`;

exports[`Valdiation Invalid should validate a Lambda with invalid config functionnArn 1`] = `
"/authentication/config/functionArn: must be a string or a CloudFormation intrinsic function
/authentication/config: must specify functionName, functionArn or function (all exclusives)
/authentication/config/functionArn: must be string
/authentication/config/identityValidationExpression: must be string
/authentication/config/authorizerResultTtlInSeconds: must be number"
`;

exports[`Valdiation Invalid should validate a Lambda with invalid config: both functionName and functionnArn are set 1`] = `"/authentication/config: must specify functionName, functionArn or function (all exclusives)"`;

exports[`Valdiation Invalid should validate a Lambda with invalid functionName and functionVersion 1`] = `
"/authentication/config: must specify functionName, functionArn or function (all exclusives)
/authentication/config/functionName: must be string
/authentication/config/identityValidationExpression: must be string
/authentication/config/authorizerResultTtlInSeconds: must be number"
`;

exports[`Valdiation Invalid should validate a Lambda with missing config 1`] = `"/authentication: must have required property 'config'"`;

exports[`Valdiation Invalid should validate a OIDC with empty config 1`] = `"/authentication/config: must have required property 'issuer'"`;

exports[`Valdiation Invalid should validate a OIDC with invalid config 1`] = `
"/authentication/config/issuer: must be string
/authentication/config/clientId: must be string
/authentication/config/iatTTL: must be number
/authentication/config/authTTL: must be number"
`;

exports[`Valdiation Invalid should validate a OIDC with missing config 1`] = `"/authentication: must have required property 'config'"`;
